### PR TITLE
PWN-5319 - Solend withdraw and deposit requests

### PR DESCRIPTION
**Pull request by 1sakov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-5319

## Description of Work

Making requests for deposit and withdraw flows

## Screenshots (Optional)

Screenshot #1 | Screenshot #2 | Screenshot #3
:------------:|:-------------:|:------------:
[Paste Pic 1] | [Paste Pic 2] | [Paste Pic 3]

### CODE REVIEW

**Comment by 1sakov** [comment]
Main focus for `SolendDepositInteractor` and `SolendWithdrawInteractor` classes guys

File: app/src/main/java/org/p2p/wallet/AppModule.kt [hunk 1]
```diff
@@ -67,7 +67,7 @@ object AppModule {
                 TransactionModule.create(),
                 AnalyticsModule.create(),
                 FeatureTogglesModule.create(),
-                P2PSdkModule.create(),
+                AppSdkModule.create(),
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/AppModule.kt:70]: KeyAppSdkModule mb?

File: app/src/main/java/org/p2p/wallet/auth/interactor/CreateWalletInteractor.kt [hunk 2]
```diff
@@ -96,7 +96,7 @@ class CreateWalletInteractor(
 
     private fun finishAuthFlow() {
         signUpFlowDataRepository.userAccount?.also {
-            tokenKeyProvider.secretKey = it.secretKey
+            tokenKeyProvider.keypair = it.keypair
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/CreateWalletInteractor.kt:99]: keypair -> keyPair

File: app/src/main/java/org/p2p/wallet/infrastructure/network/provider/TokenKeyProvider.kt [hunk 3]
```diff
@@ -43,7 +43,7 @@ class TokenKeyProvider(
         }
     }
 
-    var secretKey: ByteArray = getSecretKeyFromStorage()
+    var keypair: ByteArray = getSecretKeyFromStorage()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/provider/TokenKeyProvider.kt:46]: getSecretKeyFromStorage  -> getKeyPairFromStorage

why keypair anyway?

File: app/src/main/java/org/p2p/wallet/root/RootActivity.kt [hunk 4]
```diff
@@ -73,7 +73,7 @@ class RootActivity : BaseMvpActivity<RootContract.View, RootContract.Presenter>(
         deeplinksManager.mainFragmentManager = supportFragmentManager
         handleDeeplink()
 
-        registerNetworkObserver()
+//        registerNetworkObserver()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/root/RootActivity.kt:76]: rollback?

File: app/src/main/java/org/p2p/wallet/rpc/repository/solana/RpcSolanaRemoteRepository.kt [hunk 5]
```diff
@@ -26,18 +27,27 @@ class RpcSolanaRemoteRepository(
...
 
         val serializedTransaction = transaction.serialize()
         val base64Transaction = Base64Utils.encode(serializedTransaction)
-        val params = arrayListOf(base64Transaction, RpcSendTransactionConfig())
+        val params = arrayListOf(base64Transaction, RpcSendTransactionConfig(encoding))
 
         return api.sendTransaction(RpcRequest(method = "sendTransaction", params = params)).result
     }
 
+    override suspend fun sendTransaction(serializedTransaction: String, encoding: Encoding): String {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/rpc/repository/solana/RpcSolanaRemoteRepository.kt:46]: i haven't used encoding here, should i add one?

File: app/src/main/java/org/p2p/wallet/solend/interactor/SolendDepositInteractor.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            relayProgramId = relayProgramId,
+            token = token,
+            depositAmount = amountInLamports,
+            remainingFreeTransactionsCount = remainingFreeTransactionsCount,
+            lendingMarketAddress = null,
+            blockhash = recentBlockhash,
+            payFeeWithRelay = hasFreeTransactions,
+            feePayerToken = null,
+            realFeePayerAddress = realFeePayerAddress,
+        ) ?: error("Error occurred while creating a withdraw transaction")
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/solend/interactor/SolendDepositInteractor.kt:61]: maybe it's repository that needs to throw this error then?

File: app/src/main/java/org/p2p/wallet/solend/interactor/SolendWithdrawInteractor.kt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+class SolendWithdrawInteractor(
+    private val rpcSolanaRepository: RpcSolanaRepository,
+    private val rpcBlockhashRepository: RpcBlockhashRepository,
+    private val solendRepository: SolendRepository,
+    private val relayRepository: RelayRepository,
+    private val feeRelayerAccountInteractor: FeeRelayerAccountInteractor,
+    private val tokenKeyProvider: TokenKeyProvider
+) {
+
+    suspend fun withdraw(token: SolendDepositToken, amountInLamports: BigInteger): String {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/solend/interactor/SolendWithdrawInteractor.kt:26]: withdraw -> withdrawToken

File: app/src/main/java/org/p2p/wallet/solend/ui/withdraw/SolendWithdrawFragment.kt [hunk 8]
```diff
@@ -230,10 +194,7 @@ class SolendWithdrawFragment :
                         SolendTransactionDetailsState.Withdraw(
                             // TODO PWN-5319 add real data!!
                             TransactionDetailsViewData(
-                                amount = amount,
-                                transferFee = null,
-                                fee = "0.05 USDC (~\$0.5)",
-                                total = amount
+                                amount = amount, transferFee = null, fee = "0.05 USDC (~\$0.5)", total = amount
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/solend/ui/withdraw/SolendWithdrawFragment.kt:197]: on separate lines please

File: solanaj/src/main/java/org/p2p/solanaj/core/Account.kt [hunk 9]
```diff
@@ -20,18 +16,29 @@ class Account {
...
     }
 
     private constructor(keyPair: TweetNaclFast.Signature.KeyPair) {
         this.keyPair = keyPair
     }
 
+    /**
+     * Returns an encoded [keyPair] in Base58
+     * */
+    fun getEncodedKeyPair(): String =
```
Comment (gslevinkov) [inline solanaj/src/main/java/org/p2p/solanaj/core/Account.kt:30]: you can return Base58String here and remove KDoc. code is more documented this way

File: solanaj/src/main/java/org/p2p/solanaj/core/Account.kt [hunk 10]
```diff
@@ -20,18 +16,29 @@ class Account {
...
+        Base58Utils.encode(keypair)
+
     val publicKey: PublicKey
         get() = PublicKey(keyPair.publicKey)
 
-    val secretKey: ByteArray
+    /**
+     * The secret key is 32 byte array
+     *
+     * The keypair is 64 byte array which consists of encoded secret key and public key
```
Comment (gslevinkov) [inline solanaj/src/main/java/org/p2p/solanaj/core/Account.kt:39]: > the keypair is 64 byte array which consists of encoded secret key and public key.  

but you still return only secretKey from this keypair, not keypair itself

File: solanaj/src/main/java/org/p2p/solanaj/core/Message.java [hunk 11]
```diff
@@ -163,10 +157,97 @@ public byte[] serialize() {
         return out.array();
     }
 
+    public static Message deserialize(byte[] byteArray) {
```
Comment (gslevinkov) [inline solanaj/src/main/java/org/p2p/solanaj/core/Message.java:160]: oh daymn, huge respect for this

byteArray -> messageInBytes

File: solanaj/src/main/java/org/p2p/solanaj/core/TransactionParser.kt [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+@Suppress("Unused")
+object TransactionParser {
+
+    fun fromByteArray(byteArray: ByteArray): Transaction {
+        val signatures = mutableListOf<String>()
+        val signatureLength = ShortvecEncoding.decodeLength(byteArray)
+
+        var updatedByteArray = byteArray
+        Log.d("###", "parser $signatureLength")
```
Comment (gslevinkov) [inline solanaj/src/main/java/org/p2p/solanaj/core/TransactionParser.kt:17]: we have SolanajTimberLogger

File: solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package org.p2p.solanaj.model.types
+
+import com.google.gson.annotations.SerializedName
+
+enum class Encoding(
```
Comment (gslevinkov) [inline solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt:5]: Encoding for what?
RequestEncoding?
RpcParamsEncoding?

File: solanaj/src/main/java/org/p2p/solanaj/utils/TweetNaclFast.java [hunk 14]
```diff
@@ -820,18 +820,18 @@ public static KeyPair keyPair() {
             crypto_sign_keypair(kp.getPublicKey(), kp.getSecretKey(), false);
             return kp;
         }
-        public static KeyPair keyPair_fromSecretKey(byte [] secretKey) {
+        public static KeyPair keyPair_fromSecretKey(byte [] keypair) {
```
Comment (gslevinkov) [inline solanaj/src/main/java/org/p2p/solanaj/utils/TweetNaclFast.java:823]: fromSecretKey but you pass keypair, not ok

File: app/src/main/java/org/p2p/wallet/solend/interactor/SolendWithdrawInteractor.kt (hunk 7 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/solend/interactor/SolendWithdrawInteractor.kt:26]: I don't really want to rename the obvious name.
You  can see that this interactor is about Solend withdraw and you see that there is a token in params. 

So why should I duplicate a token in the method name? 
One of my proposal is to make the naming as shorter as possible

File: solanaj/src/main/java/org/p2p/solanaj/core/Account.kt (hunk 10 above)
Comment (1sakov) [reply solanaj/src/main/java/org/p2p/solanaj/core/Account.kt:39]: This is the description of keypair, not of the method itself and what it returns

File: solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt (hunk 13 above)
Comment (1sakov) [reply solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt:5]: This encoding is for everything

File: solanaj/src/main/java/org/p2p/solanaj/core/Account.kt (hunk 10 above)
Comment (1sakov) [reply solanaj/src/main/java/org/p2p/solanaj/core/Account.kt:39]: This is the description of keypair, not of the method itself and what it returns

File: solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt (hunk 13 above)
Comment (1sakov) [reply solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt:5]: This encoding is for everything

Comment (1sakov) [reply solanaj/src/main/java/org/p2p/solanaj/model/types/Encoding.kt:5]: This encoding is for everything.
Again, let's make it simple and short

File: app/src/main/java/org/p2p/wallet/infrastructure/network/provider/TokenKeyProvider.kt (hunk 3 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/provider/TokenKeyProvider.kt]: We have a big problem which I've figured out.
Will talk in chat

File: solanaj/src/main/java/org/p2p/solanaj/utils/TweetNaclFast.java (hunk 14 above)
Comment (1sakov) [reply solanaj/src/main/java/org/p2p/solanaj/utils/TweetNaclFast.java:823]: too much to refactor in this PR. 
This should be in tech debt

**Comment by 1sakov** [comment]
Sorry for BIG PR, it was necessary

**Review by gslevinkov — APPROVED** [review]

**Pull request merged**